File: streamflowSignatures/rpkg/tests/testthat/test-drought_metrics.R
# Streamflow drought (Phase 5). Cross-checked against canonical Julia on
# 2026-08-26: 105 values over three fixtures (seasonal, intermittent, gapped),
# 0 mismatches, worst relative difference 3.6e-14.

mk_dr <- function(f, n_years = 25, skip = NULL) {
  rows <- list()
  for (wy in 1996:(1995 + n_years)) {
    d0 <- as.Date(paste0(wy - 1, "-10-01")); d1 <- as.Date(paste0(wy, "-09-30"))
    nd <- as.integer(d1 - d0) + 1L
    keep <- rep(TRUE, nd)
    if (!is.null(skip)) for (i in 1:nd) keep[i] <- !skip(i, wy)
    df <- data.frame(date = d0 + (0:(nd - 1L)), water_year = wy, dowy = 1:nd,
                     Q = vapply(1:nd, function(i) f(i, wy), numeric(1)))
    rows[[length(rows) + 1L]] <- df[keep, , drop = FALSE]
  }
  do.call(rbind, rows)
}
PCTS <- c(2, 5, 10, 20, 30)
DUR <- paste0("drought_duration_fixed_p", PCTS)
DEF <- paste0("drought_deficit_fixed_p", PCTS)
THR <- paste0("drought_threshold_fixed_p", PCTS)

test_that("weibull_quantile: hand-derived interpolation and range policy", {
  x <- 1:9                                    # n = 9, positions i/10
  expect_equal(weibull_quantile(x, 0.5), 5)   # h = 5 -> the 5th value
  expect_equal(weibull_quantile(x, 0.25), 2.5)
  expect_true(is.na(weibull_quantile(x, 0.05)))               # below 1/(n+1)
  expect_equal(weibull_quantile(x, 0.05, below_range = "clamp"), 1)
  expect_equal(weibull_quantile(x, 0.99), 9)                  # above n/(n+1)
  expect_equal(weibull_quantile(c(1, NA, 2, 3, NA), 0.5), 2)  # NAs dropped
})

test_that("smoothing shrinks at edges and NEVER averages across a gap", {
  d <- as.Date("2000-01-01") + 0:9
  out <- smooth_daily_flow(d, 1:10, window = 7, alignment = "center", min_valid = 1)
  expect_equal(out[4], 4)      # interior: mean of days 1..7
  expect_equal(out[1], 2.5)    # edge: window shrinks to days 1..4

  dg <- as.Date("2000-01-01") + c(0, 1, 2, 3, 4, 6, 7, 8, 9, 10)   # 1-day hole
  og <- smooth_daily_flow(dg, c(rep(1, 5), rep(100, 5)), window = 7,
                          alignment = "center", min_valid = 1)
  expect_equal(og[1:5], rep(1, 5))      # untouched by the 100s across the gap
  expect_equal(og[6:10], rep(100, 5))
})

test_that("smoothing: min_valid and duplicate dates", {
  d2 <- as.Date("2000-01-01") + 0:1
  expect_true(all(is.na(smooth_daily_flow(d2, c(1, 2), window = 7,
                                          alignment = "center", min_valid = 4))))
  ddup <- as.Date("2000-01-01") + c(0, 1, 1, 2)   # duplicate breaks the run
  od <- smooth_daily_flow(ddup, c(1, 2, 3, 4), window = 3,
                          alignment = "center", min_valid = 1)
  expect_equal(od[1], 1.5)
  expect_false(any(is.na(od)))
})

test_that("seasonal gage: all metrics, monotone levels, ordered thresholds", {
  r <- calculate_drought_metrics(
    mk_dr(function(i, wy) 2 + 1.8*sin(2*pi*(i-30)/365) + 0.3*((i*7919) %% 11)/11),
    trend_completeness = 0.6, decade_completeness = 0.8, min_values_for_stats = 20)
  for (m in c(DUR, DEF)) expect_true(paste0(m, "_mean") %in% names(r), info = m)
  durs <- vapply(DUR, function(m) r[[paste0(m, "_mean")]], numeric(1))
  defs <- vapply(DEF, function(m) r[[paste0(m, "_mean")]], numeric(1))
  thr  <- vapply(THR, function(s) r[[s]], numeric(1))
  expect_false(is.unsorted(durs))     # non-decreasing in severity level
  expect_false(is.unsorted(defs))
  expect_false(is.unsorted(thr))
  # weak construction anchor: a level-p threshold sits below ~p% of days
  expect_gt(r$drought_duration_fixed_p10_mean, 30)
  expect_lt(r$drought_duration_fixed_p10_mean, 43)
})

test_that("intermittent gage: zero thresholds + STRICT < give zero duration", {
  r <- calculate_drought_metrics(
    mk_dr(function(i, wy) if (i >= 150 && i <= 300) 0 else 1 + ((i*104729) %% 7)/7),
    trend_completeness = 0.6, decade_completeness = 0.8)
  expect_equal(r$drought_threshold_fixed_p2, 0)
  expect_equal(r$drought_duration_fixed_p2_mean, 0)
  expect_equal(r$drought_deficit_fixed_p2_mean, 0)
})

test_that("a low-flow block straddling Sep 30 -> Oct 1 is SPLIT across years", {
  # Codex MAJOR-1 on the Julia original: conservation alone would pass under a
  # uniform water-year shift, so attribution must be pinned explicitly.
  lo_start <- as.Date("2005-09-25"); lo_end <- as.Date("2005-10-07")
  f <- function(i, wy) {
    d <- as.Date(paste0(wy - 1, "-10-01")) + (i - 1L)
    if (d >= lo_start && d <= lo_end) 0 else 10
  }
  co <- annual_collector()
  calculate_drought_metrics(mk_dr(f), collector = co)
  d <- collector_drain(co)
  dur <- d[d$signature == "drought_duration_fixed_p30", ]
  v05 <- dur$value[dur$water_year == 2005]; v06 <- dur$value[dur$water_year == 2006]
  expect_gt(v05, 0); expect_gt(v06, 0)                    # SPLIT, not lumped
  expect_gte(v05 + v06, 13)
  others <- dur$value[!dur$water_year %in% c(2005, 2006)]
  expect_true(all(others == 0))                            # uniform-shift guard
})

test_that("schema contract on empty / missing-column / short records", {
  cp <- list(start_year=1900, end_year=2100, min_total_obs=20, min_segment_obs=10)
  empty <- data.frame(date = as.Date(character(0)), water_year = integer(0), Q = numeric(0))
  r <- calculate_drought_metrics(empty, changepoint = cp)
  for (m in c(DUR, DEF)) {
    expect_true(is.na(r[[paste0(m, "_mean")]]), info = m)
    expect_true(paste0(m, "_pettitt_cp_year") %in% names(r), info = m)
  }
  for (s in THR) expect_true(is.na(r[[s]]), info = s)

  r2 <- suppressWarnings(calculate_drought_metrics(data.frame(Q = c(1, 2)), changepoint = cp))
  for (m in c(DUR, DEF)) expect_true(is.na(r2[[paste0(m, "_mean")]]), info = m)

  # fewer than min_years_for_threshold -> NA thresholds, hence NA metrics
  r3 <- calculate_drought_metrics(mk_dr(function(i, wy) 1 + 0.5*sin(i/30), n_years = 5))
  for (s in THR) expect_true(is.na(r3[[s]]), info = s)
})

test_that("END-TO-END: drought runs on a preprocess_daily_data() output", {
  # Regression for 2026-08-26: rpkg's preprocessor renames `date` -> `Date`, so
  # drought (which checks the Julia/Python lowercase name) silently produced an
  # all-NA family for every real gage. The other tests in this file build frames
  # with `date` directly and cannot catch that; this one goes through the actual
  # pipeline the benchmark runner uses.
  set.seed(9)
  dates <- seq(as.Date("1992-10-01"), as.Date("2025-09-30"), by = "day")
  q <- 2 + 1.5 * sin(2 * pi * as.integer(format(dates, "%j")) / 365) + runif(length(dates)) * 0.2
  gd <- add_water_year_columns(data.table::data.table(gage_id = "T", date = dates, Q = q))
  pp <- preprocess_daily_data(gd)
  expect_gte(length(pp$valid_years), 20)

  r <- calculate_drought_metrics(pp$data, trend_completeness = 0.6,
                                 decade_completeness = 0.8, min_values_for_stats = 20)
  expect_false(is.na(r$drought_threshold_fixed_p10))
  expect_false(is.na(r$drought_duration_fixed_p10_mean))
  expect_gt(r$drought_duration_fixed_p10_mean, 0)

  # ...and through the orchestrator, the way the runner calls it
  s <- calculate_all_signatures(pp$data, has_climate = FALSE,
                                trend_completeness = 0.6, decade_completeness = 0.8,
                                min_values_for_stats = 20, gage_id = "T")
  expect_false(is.na(s$drought_duration_fixed_p10_mean))
})

test_that("smoothing accumulates SEQUENTIALLY in double, matching julia/src/drought.jl", {
  # R's mean() uses a long-double accumulator plus a correction pass, so it can
  # return a different last bit than Julia's `s += v` loop. Harmless alone, but
  # the fixed thresholds are percentiles OF this series: when a threshold lands
  # on a flow plateau, a last-bit shift flips every plateau day at once through
  # the strict `<`. Measured 2026-08-26 on gage 01589795 — 4,513 of 10,227
  # smoothed values differed (max 3.6e-15) and WY2002's p5 drought duration read
  # 116 in Julia vs 60 here. After this change the series is BIT-IDENTICAL to
  # Julia across all 10,227 days and the duration agrees.
  v <- c(0.098890929785557094, 0.039774545328691603, 0.011569777876138687,
         0.0069748678710311656, 0.024374939058907332, 0.079201042582280945,
         0.034006235282868148)
  seq_sum <- 0
  for (x in v) seq_sum <- seq_sum + x
  expect_false(identical(seq_sum / 7, mean(v)))   # the fixture really discriminates

  dates <- seq(as.Date("2000-01-01"), by = "day", length.out = 7)
  out <- smooth_daily_flow(dates, v, window = 7L, alignment = "center", min_valid = 1L)
  # the centre day sees the whole window
  expect_identical(out[4], seq_sum / 7)
  expect_false(identical(out[4], mean(v)))
})

test_that("zero-row input returns the canonical key set for every sparse family", {
  # Regression for 2026-08-26: these functions built a 0-row frame and then
  # assigned length-1 NA columns into it, raising an error that the orchestrator
  # swallowed — silently dropping the family for that gage.
  d <- data.table::data.table(
    gage_id = "T", date = seq(as.Date("1992-10-01"), as.Date("1995-09-30"), by = "day"))
  d[, Q := 1.5]
  d <- add_water_year_columns(d)

  for (fn in c("analyze_flashiness_trends", "analyze_flow_timing_trends",
               "analyze_baseflow_indices")) {
    f <- get(fn, envir = asNamespace("streamflowsignatures"))
    rz <- f(d[0]); rn <- f(d)
    expect_identical(sort(names(rz)), sort(names(rn)), info = fn)
    expect_true(all(is.na(unlist(rz))), info = fn)
  }
})

File: streamflowSignatures/python/tests/test_drought_metrics.py
"""
Tests for the 10 streamflow drought metrics + 5 threshold scalars (Phase 5 of
the port campaign; port of julia/test/test_drought_metrics.jl).

Cross-checked against canonical Julia on 2026-08-25: 105 values across three
fixtures (seasonal low-flow, intermittent/zero-threshold, and a gapped record)
— 0 mismatches, worst relative difference 6.0e-15.

Includes the boundary-attribution fixture Codex required of the Julia original
(MAJOR-1): a conservation check alone would pass under a uniform water-year
shift, so a low-flow block straddling Sep 30 -> Oct 1 is split explicitly.

Run with: pytest tests/test_drought_metrics.py -v
"""

from datetime import date, timedelta

import numpy as np
import pandas as pd
import pytest

from streamflow_signatures.drought import (
    calculate_drought_metrics, weibull_quantile, smooth_daily_flow,
    DROUGHT_METRICS, DROUGHT_DURATION_METRICS, DROUGHT_DEFICIT_METRICS,
    DROUGHT_THRESHOLD_SCALARS,
)
from streamflow_signatures import AnnualCollector

CP_CONFIG = {"start_year": 1900, "end_year": 2100,
             "min_total_obs": 20, "min_segment_obs": 10}


def make_gage(q_of, n_years=25, start=1996, skip=None):
    rows = []
    for wy in range(start, start + n_years):
        d0, d1 = date(wy - 1, 10, 1), date(wy, 9, 30)
        nd = (d1 - d0).days + 1
        for i in range(1, nd + 1):
            if skip is not None and skip(i, wy):
                continue
            rows.append((pd.Timestamp(d0 + timedelta(days=i - 1)), wy, i, float(q_of(i, wy))))
    return pd.DataFrame(rows, columns=["date", "water_year", "dowy", "Q"])


class TestWeibullQuantile:
    def test_hand_derived_interpolation(self):
        """p_i = i/(n+1): for n=9, p=0.5 -> h=5 -> exactly the 5th value."""
        x = np.arange(1.0, 10.0)                    # 1..9, n = 9
        assert weibull_quantile(x, 0.5) == pytest.approx(5.0)
        # p = 0.25 -> h = 2.5 -> midpoint of the 2nd and 3rd values
        assert weibull_quantile(x, 0.25) == pytest.approx(2.5)

    def test_below_plotting_range_is_na_not_clamped(self):
        """Below 1/(n+1) the sample cannot resolve the probability: NaN, NOT the
        minimum (a project policy layered on Hyndman-Fan type 6)."""
        x = np.arange(1.0, 10.0)                    # smallest position = 0.1
        assert np.isnan(weibull_quantile(x, 0.05))
        assert weibull_quantile(x, 0.05, below_range="clamp") == pytest.approx(1.0)

    def test_above_range_returns_max(self):
        x = np.arange(1.0, 10.0)
        assert weibull_quantile(x, 0.99) == pytest.approx(9.0)

    def test_nans_dropped_and_empty(self):
        x = np.array([1.0, np.nan, 2.0, 3.0, np.nan])
        assert weibull_quantile(x, 0.5) == pytest.approx(2.0)
        assert np.isnan(weibull_quantile(np.array([np.nan]), 0.5))


class TestSmoothing:
    def _dates(self, n, start="2000-01-01"):
        return pd.date_range(start, periods=n, freq="D").to_numpy()

    def test_centered_window_and_edge_shrink(self):
        q = np.arange(1.0, 11.0)
        out = smooth_daily_flow(self._dates(10), q, window=7, alignment="center",
                                min_valid=1)
        # interior day 4 (0-based 3): mean of days 1..7 = 4
        assert out[3] == pytest.approx(4.0)
        # first day: window shrinks to days 1..4 -> mean 2.5
        assert out[0] == pytest.approx(2.5)

    def test_min_valid_yields_nan(self):
        q = np.arange(1.0, 11.0)
        out = smooth_daily_flow(self._dates(10), q, window=7, alignment="center",
                                min_valid=4)
        assert not np.isnan(out[3])
        # a 2-day run cannot reach min_valid=4
        short = smooth_daily_flow(self._dates(2), np.array([1.0, 2.0]), window=7,
                                  alignment="center", min_valid=4)
        assert np.isnan(short).all()

    def test_never_averages_across_a_gap(self):
        """Two runs separated by a missing day must not blend."""
        dates = np.array([np.datetime64("2000-01-01") + np.timedelta64(d, "D")
                          for d in [0, 1, 2, 3, 4, 6, 7, 8, 9, 10]])
        q = np.array([1.0] * 5 + [100.0] * 5)
        out = smooth_daily_flow(dates, q, window=7, alignment="center", min_valid=1)
        assert out[:5] == pytest.approx([1.0] * 5)      # untouched by the 100s
        assert out[5:] == pytest.approx([100.0] * 5)

    def test_duplicate_date_breaks_the_run(self):
        dates = np.array([np.datetime64("2000-01-01") + np.timedelta64(d, "D")
                          for d in [0, 1, 1, 2]])
        out = smooth_daily_flow(dates, np.array([1.0, 2.0, 3.0, 4.0]), window=3,
                                alignment="center", min_valid=1)
        assert np.isfinite(out).all()                   # runs are just shorter
        assert out[0] == pytest.approx(1.5)             # run = [1.0, 2.0]

    def test_length_mismatch_raises(self):
        with pytest.raises(ValueError):
            smooth_daily_flow(self._dates(3), np.array([1.0, 2.0]))


class TestSeasonalGage:
    @pytest.fixture(scope="class")
    def r(self):
        gage = make_gage(lambda i, wy: 2.0 + 1.8*np.sin(2*np.pi*(i-30)/365)
                         + 0.3*((i*7919) % 11)/11)
        return calculate_drought_metrics(gage, trend_completeness=0.6,
                                         decade_completeness=0.8,
                                         min_values_for_stats=20)

    def test_all_metrics_and_scalars_present(self, r):
        for m in DROUGHT_METRICS:
            assert f"{m}_mean" in r
        for s in DROUGHT_THRESHOLD_SCALARS:
            assert s in r and np.isfinite(r[s])

    def test_level_monotonicity(self, r):
        """Duration and deficit are non-decreasing in the severity level by
        construction (a higher percentile threshold catches more days)."""
        durs = [r[f"{m}_mean"] for m in DROUGHT_DURATION_METRICS]
        defs = [r[f"{m}_mean"] for m in DROUGHT_DEFICIT_METRICS]
        assert durs == sorted(durs)
        assert defs == sorted(defs)

    def test_thresholds_ordered(self, r):
        thr = [r[s] for s in DROUGHT_THRESHOLD_SCALARS]
        assert thr == sorted(thr)

    def test_weak_construction_anchor(self, r):
        """A level-p threshold sits below ~p% of all days by construction, so
        duration_p10 averages ~36.5 d/yr. Near-circular — a sanity anchor only,
        as documented."""
        assert 30.0 < r["drought_duration_fixed_p10_mean"] < 43.0


class TestIntermittentGage:
    def test_zero_threshold_strict_comparison(self):
        """Where most days are zero flow the low thresholds are exactly 0, and
        the STRICT `<` then reports 0 duration/deficit — the documented
        interpretation trap, pinned here."""
        gage = make_gage(lambda i, wy: 0.0 if 150 <= i <= 300
                         else 1.0 + ((i*104729) % 7)/7)
        r = calculate_drought_metrics(gage, trend_completeness=0.6,
                                      decade_completeness=0.8)
        assert r["drought_threshold_fixed_p2"] == pytest.approx(0.0)
        assert r["drought_duration_fixed_p2_mean"] == pytest.approx(0.0)
        assert r["drought_deficit_fixed_p2_mean"] == pytest.approx(0.0)


class TestBoundaryAttribution:
    """Codex MAJOR-1 on the Julia original: a conservation check alone would
    pass under a UNIFORM water-year shift, so attribution must be pinned
    explicitly with a low-flow block straddling Sep 30 -> Oct 1."""

    def test_block_split_across_the_water_year_boundary(self):
        # High baseline everywhere except a 13-day low block spanning
        # 2005-09-25..2005-10-07 — i.e. 6 days in WY2005 and 7 days in WY2006.
        lo_start, lo_end = date(2005, 9, 25), date(2005, 10, 7)

        def q_of(i, wy):
            d = date(wy - 1, 10, 1) + timedelta(days=i - 1)
            return 0.0 if lo_start <= d <= lo_end else 10.0

        gage = make_gage(q_of, n_years=25)
        c = AnnualCollector()
        calculate_drought_metrics(gage, collector=c)
        dur = {y: v for s, y, v in zip(c.signature, c.water_year, c.value)
               if s == "drought_duration_fixed_p30"}
        # The block must be SPLIT, not attributed wholly to either year
        assert dur[2005] > 0 and dur[2006] > 0
        assert dur[2005] + dur[2006] >= 13          # smoothing widens the dip
        # ...and no other year sees drought (uniform-shift guard)
        others = [v for y, v in dur.items() if y not in (2005, 2006)]
        assert all(v == 0 for v in others)


class TestSchemaContract:
    def test_empty_frame_emits_full_key_set(self):
        empty = pd.DataFrame({"date": pd.Series([], dtype="datetime64[ns]"),
                              "water_year": pd.Series([], dtype="int64"),
                              "Q": pd.Series([], dtype="float64")})
        r = calculate_drought_metrics(empty, changepoint=CP_CONFIG)
        for m in DROUGHT_METRICS:
            assert np.isnan(r[f"{m}_mean"])
            assert f"{m}_pettitt_cp_year" in r
        for s in DROUGHT_THRESHOLD_SCALARS:
            assert np.isnan(r[s])

    def test_missing_columns_emit_full_key_set(self):
        r = calculate_drought_metrics(pd.DataFrame({"Q": [1.0, 2.0]}),
                                      changepoint=CP_CONFIG)
        for m in DROUGHT_METRICS:
            assert np.isnan(r[f"{m}_mean"])

    def test_short_record_gets_nan_thresholds(self):
        """Fewer than min_years_for_threshold qualifying years -> NaN thresholds
        (and hence NaN metrics) rather than an unstable low tail."""
        gage = make_gage(lambda i, wy: 1.0 + 0.5*np.sin(i/30), n_years=5)
        r = calculate_drought_metrics(gage)
        for s in DROUGHT_THRESHOLD_SCALARS:
            assert np.isnan(r[s])
        for m in DROUGHT_METRICS:
            assert np.isnan(r[f"{m}_mean"])
